### PR TITLE
AK+LibCrypto+LibTLS+Kernel: Switch the Cipher::Mode interface to use Span

### PR DESCRIPTION
**Pull request by alimpfard**
Following #3089, this gets rid of more ByteBuffer overhead, bringing the runtime of `disasm /bin/id` down to 2.5 seconds (best of 3)

and also (technically) doubles(-ish) the speed of all AES operations performed through the Mode API.

### CODE REVIEW

File: Kernel/Random.h [hunk 1]
```diff
@@ -68,12 +68,13 @@ class FortunaPRNG {
 
         typename CipherType::CTRMode cipher(m_key, KeySize);
 
-        auto wrapped_buffer = ByteBuffer::wrap(buffer, n);
-        m_counter = cipher.key_stream(wrapped_buffer, m_counter).value();
+        Span<u8> buffer_span { buffer, n };
```
Comment (nico) [inline Kernel/Random.h:71]: We have the `Bytes` typedef for `Span<u8>`.

File: Libraries/LibCrypto/Authentication/HMAC.h [hunk 2]
```diff
@@ -68,8 +68,11 @@ class HMAC {
         m_inner_hasher.update(message, length);
     }
 
+    TagType process(const Span<const u8>& span) { return process(span.data(), span.size()); }
```
Comment (nico) [inline Libraries/LibCrypto/Authentication/HMAC.h:71]: We have `ReadonlyBytes` for `Span<const u8>`.

File: Kernel/Random.h (hunk 1 above)
Comment (alimpfard) [reply Kernel/Random.h:71]: Ah, I didn't see that.
will change.

File: Userland/test-crypto.cpp [hunk 3]
```diff
@@ -99,7 +99,7 @@ int crc32_tests();
 
 // stop listing tests
 
-void print_buffer(const ByteBuffer& buffer, int split)
+void print_buffer(const Span<const u8>& buffer, int split)
```
Comment (awesomekling) [inline Userland/test-crypto.cpp:102]: Seems like this should be `ReadonlyBytes`

File: Libraries/LibTLS/Record.cpp [hunk 4]
```diff
@@ -77,12 +77,12 @@ void TLSv12::update_packet(ByteBuffer& packet)
 
             if (m_context.crypto.created == 1) {
                 // `buffer' will continue to be encrypted
-                auto buffer = ByteBuffer::create_zeroed(length);
+                auto buffer = ByteBuffer::create_uninitialized(length);
```
Comment (awesomekling) [inline Libraries/LibTLS/Record.cpp:80]: Hmm?

File: Libraries/LibTLS/Record.cpp [hunk 5]
```diff
@@ -77,12 +77,12 @@ void TLSv12::update_packet(ByteBuffer& packet)
...
                 size_t buffer_position = 0;
                 auto iv_size = iv_length();
 
                 // We need enough space for a header, iv_length bytes of IV and whatever the packet contains
-                auto ct = ByteBuffer::create_zeroed(length + header_size + iv_size);
+                auto ct = ByteBuffer::create_uninitialized(length + header_size + iv_size);
```
Comment (awesomekling) [inline Libraries/LibTLS/Record.cpp:85]: Hmm?

File: Libraries/LibTLS/Record.cpp (hunk 4 above)
Comment (alimpfard) [reply Libraries/LibTLS/Record.cpp:80]: There's no need for that to be zeroed, it's initialised right there

`buffer`:
- L91: 0:(`packet.size() - header_size`)
- L98: (`packet.size() - header_size`):(`packet.size() - header_size + mac_size`)
- L102: (`packet.size() - header_size + mac_size`):`buffer.size()`
- (asserted at L103)

`ct`:
- L88: 0:(`header_size - 2`)
- L123: (`header_size - 2`):(`header_size`)
- L111: (`header_size`):(`header_size + iv_size`)
- L117: (`header_size + iv_size`):(`header_size + iv_size + length`)
- (asserted at L113)

Comment (awesomekling) [reply Libraries/LibTLS/Record.cpp:80]: I believe you, just seems like a sneaky unrelated change not mentioned in commit message or anywhere :)

File: Userland/test-crypto.cpp (hunk 3 above)
Comment (alimpfard) [reply Userland/test-crypto.cpp]: _/me patches stuff and doesn't add it before amending_

File: Libraries/LibTLS/Record.cpp (hunk 4 above)
Comment (alimpfard) [reply Libraries/LibTLS/Record.cpp:80]: I'll split it up

(that also served to prove it to myself...)

**Comment by awesomekling** [comment]
:star:

**Pull request merged**